### PR TITLE
webgpu: Optimize softmax

### PR DESCRIPTION
**Pull request by qjia7**
PERF

Previously, the softmax was implemented by the naive method, which
includes five (max, sub, exp, sum, div) basic kernels. Even we fused sub
and exp, it sill needs 4 compute shaders. This PR utilizes shared memory
to implement everything in one shader and avoids the intermediate
buffers reading/writing, which greatly reduces the I/O accessing time.

softmax time in stable diffusion: 74ms -> 16ms

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7588)

### CODE REVIEW

File: tfjs-backend-webgpu/src/setup_test.ts [hunk 1]
```diff
@@ -66,55 +66,63 @@ const TEST_FILTERS: TestFilter[] = [
...
     excludes: [
-      'log propagates NaNs', // Failing on MacOS
+      'log propagates NaNs',  // Failing on MacOS
+    ],
+  },
+
+  {
+    startsWith: 'softmax ',
+    excludes: [
+      'underflow',        // Actual: NaN,NaN. Expected: 0.5,0.5.
```
Comment (hujiajie) [inline tfjs-backend-webgpu/src/setup_test.ts:124]: Why is this failing?

File: tfjs-backend-webgpu/src/softmax_webgpu.ts [hunk 2]
```diff
@@ -16,36 +16,92 @@
...
+      let tid = i32(localId.x);
+      let cols = uniforms.outShape[1];
+
+      var threadMax = -1.0 / 1e-20;
+      for (var col = tid; col < cols; col += blockSize) {
+        let value = getLogits(row, col);
+        threadMax = max(threadMax, value);
+      }
+      if (tid < cols)
+      {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/softmax_webgpu.ts:58]: Do not use a new line for "{"?

File: tfjs-backend-webgpu/src/softmax_webgpu.ts [hunk 3]
```diff
@@ -16,36 +16,92 @@
...
+        buf[tid] = threadMax;
+      }
+      workgroupBarrier();
+
+      for (var currSize = blockSize >> 1;  currSize > 0; currSize = currSize >> 1)
+      {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/softmax_webgpu.ts:64]: Do not use a new line for "{"?

**Review by gyagp — APPROVED** [review]
LGTM with nits

File: tfjs-backend-webgpu/src/setup_test.ts (hunk 1 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: Fix it here https://github.com/tensorflow/tfjs/pull/7588/commits/7a08da099a67ca1a5a539e0ca090c4b750417c94. Thanks.

File: tfjs-backend-webgpu/src/softmax_webgpu.ts [hunk 4]
```diff
@@ -16,36 +16,89 @@
...
+    var<workgroup> buf : array<f32, ${this.workgroupSize[0]}>;
+    var<workgroup> rowMaxShared : f32;
+    var<workgroup> rowSumShared : f32;
+    const blockSize = ${this.workgroupSize[0]};
+    ${main('index')} {
+      let row = index / blockSize;
+      let tid = i32(localId.x);
+      let cols = uniforms.outShape[1];
+
+      var threadMax = -1.0 / 1e-20;
```
Comment (hujiajie) [inline tfjs-backend-webgpu/src/softmax_webgpu.ts:52]: `tf.softmax(tf.tensor1d([-2e20, -2e20]))` seems to be regressing.

File: tfjs-backend-webgpu/src/softmax_webgpu.ts [hunk 5]
```diff
@@ -16,36 +16,89 @@
...
+        rowMaxShared = max(buf[0], buf[reduceSize - 1]);
+      }
+      workgroupBarrier();
+
+      var threadSum = 0.0;
+      for (var col = tid; col < cols; col += blockSize) {
+        let subExp = exp(getLogits(row, col) - rowMaxShared);
+        threadSum += subExp;
+      }
+      if (tid < cols) {
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/softmax_webgpu.ts:80]: Because we will sum all elements of `buf` below, if cols is less than blockSize, will there be an error? I think the `if` condition could be removed here.

File: tfjs-backend-webgpu/src/kernels/Softmax.ts [hunk 6]
```diff
@@ -32,35 +29,22 @@ export function softmax(
...
-
-  backend.disposeData(maxLogit.dataId);
-  backend.disposeData(maxLogitsReshaped.dataId);
-  backend.disposeData(b.dataId);
-  backend.disposeData(sumExp.dataId);
-  backend.disposeData(sumExpReshaped.dataId);
-
-  return res;
+  const program = new SoftmaxProgram(logitsReshaped.shape);
+  const res = backend.runWebGPUProgram(program, [logitsReshaped], logits.dtype);
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/kernels/Softmax.ts:42]: I think the result elements' type is float, and (0, 1], if the dtype of `logits` is int, will there be an error?

**Review by qjia7 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7588)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp, @hujiajie, @Linchenn, and @xhcao)

___
*[`tfjs-backend-webgpu/src/softmax_webgpu.ts` line 58 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7588#gh-1169526783:-NTN-ZZo6tsANjjUKya_:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/1a7020d1c02986da8ad22c895fd4cbcaa68e6bf4/tfjs-backend-webgpu/src/softmax_webgpu.ts#L58)):*
<details><summary><i>Previously, gyagp (Yang Gu) wrote…</i></summary><blockquote>

Do not use a new line for "{"?
</blockquote></details>

Done.
___
*[`tfjs-backend-webgpu/src/softmax_webgpu.ts` line 64 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7588#gh-1169526903:-NTN-_-BChIzYaQXDSFJ:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/1a7020d1c02986da8ad22c895fd4cbcaa68e6bf4/tfjs-backend-webgpu/src/softmax_webgpu.ts#L64)):*
<details><summary><i>Previously, gyagp (Yang Gu) wrote…</i></summary><blockquote>

Do not use a new line for "{"?
</blockquote></details>

Done.
___
*[`tfjs-backend-webgpu/src/softmax_webgpu.ts` line 52 at r2](https://reviewable.io/reviews/tensorflow/tfjs/7588#gh-1170747387:-NTN-T7C4p8zgl1bDnzG:b-xo2low) ([raw file](https://github.com/tensorflow/tfjs/blob/3d7edd887e70d9d8e6ec7f8c9c9b5f329acecd80/tfjs-backend-webgpu/src/softmax_webgpu.ts#L52)):*
<details><summary><i>Previously, hujiajie (Jiajie Hu) wrote…</i></summary><blockquote>

`tf.softmax(tf.tensor1d([-2e20, -2e20]))` seems to be regressing.
</blockquote></details>

set it to -3.402823e+38f. Thanks.
___
*[`tfjs-backend-webgpu/src/softmax_webgpu.ts` line 80 at r2](https://reviewable.io/reviews/tensorflow/tfjs/7588#gh-1170774206:-NTN-_ygEFzwbwjcTjD2:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/3d7edd887e70d9d8e6ec7f8c9c9b5f329acecd80/tfjs-backend-webgpu/src/softmax_webgpu.ts#L80)):*
<details><summary><i>Previously, xhcao wrote…</i></summary><blockquote>

Because we will sum all elements of `buf` below, if cols is less than blockSize, will there be an error? I think the `if` condition could be removed here.
</blockquote></details>

Done.
___
*[`tfjs-backend-webgpu/src/kernels/Softmax.ts` line 42 at r2](https://reviewable.io/reviews/tensorflow/tfjs/7588#gh-1170775154:-NTN-o_FCSur0WoGZXDv:bwv5za) ([raw file](https://github.com/tensorflow/tfjs/blob/3d7edd887e70d9d8e6ec7f8c9c9b5f329acecd80/tfjs-backend-webgpu/src/kernels/Softmax.ts#L42)):*
<details><summary><i>Previously, xhcao wrote…</i></summary><blockquote>

I think the result elements' type is float, and (0, 1], if the dtype of `logits` is int, will there be an error?
</blockquote></details>

See https://www.tensorflow.org/api_docs/python/tf/nn/softmax, `Returns: A Tensor. Has the same type and shape as logits.` And the input logits `A non-empty Tensor. Must be one of the following types: half, float32, float64`. So it won't happen the situation you said.

**Review by xhcao — APPROVED** [review]
LGTM, thanks

File: tfjs-backend-webgpu/src/softmax_webgpu.ts [hunk 7]
```diff
@@ -16,36 +16,87 @@
...
+        let value = getLogits(row, col);
+        threadMax = max(threadMax, value);
+      }
+      if (tid < cols) {
+        buf[tid] = threadMax;
+      }
+      workgroupBarrier();
+
+      let reduceSize = min(cols, blockSize);
+      for (var currSize = reduceSize >> 1;  currSize > 0; currSize = currSize >> 1) {
```
Comment (hujiajie) [inline tfjs-backend-webgpu/src/softmax_webgpu.ts:63]: `tf.softmax(tf.tensor1d([-400, -400, 0, -400, -400, -400, -400]))` seems to be regressing.
